data/builder/directory.go

```diff
@@ -0,0 +1,122 @@
+package builder
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path"
+
+	"github.com/ipfs/go-unixfsnode/data"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/multiformats/go-multihash"
+)
+
+// https://github.com/ipfs/go-ipfs/pull/8114/files#diff-eec963b47a6e1080d9d8023b4e438e6e3591b4154f7379a7e728401d2055374aR319
+const shardSplitThreshold = 262144
+
+// https://github.com/ipfs/go-unixfs/blob/ec6bb5a4c5efdc3a5bce99151b294f663ee9c08d/io/directory.go#L29
+const defaultShardWidth = 256
+
+// BuildUnixFSRecursive returns a link pointing to the UnixFS node representing
+// the file or directory tree pointed to by `root`
+// TODO: support symlinks
+func BuildUnixFSRecursive(root string, ls *ipld.LinkSystem) (ipld.Link, uint64, error) {
+	info, err := os.Lstat(root)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	m := info.Mode()
+	switch {
+	case m.IsDir():
+		entries, err := os.ReadDir(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		lnks := make([]dagpb.PBLink, 0, len(entries))
+		for _, e := range entries {
+			lnk, sz, err := BuildUnixFSRecursive(path.Join(root, e.Name()), ls)
+			if err != nil {
+				return nil, 0, err
+			}
+			entry, err := BuildUnixFSDirectoryEntry(e.Name(), int64(sz), lnk)
+			if err != nil {
+				return nil, 0, err
+			}
+			lnks = append(lnks, entry)
+		}
+		outLnk, err := BuildUnixFSDirectory(lnks, ls)
+		return outLnk, 0, err
+	case m.Type() == fs.ModeSymlink:
+		content, err := os.Readlink(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		return BuildUnixFSSymlink(content, ls)
+	case m.IsRegular():
+		fp, err := os.Open(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		defer fp.Close()
+		return BuildUnixFSFile(fp, "", ls)
+	default:
+		return nil, 0, fmt.Errorf("cannot encode non regular file: %s", root)
+	}
+}
+
+// estimateDirSize estimates if a directory is big enough that it warrents sharding
+func estimateDirSize(entries []dagpb.PBLink) int {
```

Comment:
I might beg for slightly more docs here if possible?  ISTM that this function is super critical to whether the system converges stably or not, and that's not immediately obvious from the name of it.  It might be an estimate, but it's _deciding_, too.